### DISCUSSION TITLE
编译android平台库时，编译出错

### DISCUSSION THREAD

**Discussion by stopandjoin** [Tencent/tquic#191] [Q&A (问题和回答)] [answered]
按照文档指引，已安装所有环境。
 Ubuntu20 环境下编译， 出错：
cargo ndk -t arm64-v8a -p 21 -- build --features ffi --release
    Building arm64-v8a (aarch64-linux-android)
   Compiling proc-macro2 v1.0.79
   Compiling unicode-ident v1.0.12
   Compiling autocfg v1.1.0
   Compiling libc v0.2.153
   Compiling cc v1.0.90
   Compiling serde v1.0.197
error[E0463]: can't find crate for `core`
  |
  = note: the `aarch64-linux-android` target may not be installed
  = help: consider downloading the target with `rustup target add aarch64-linux-android`

error[E0463]: can't find crate for `compiler_builtins`

error[E0432]: unresolved import `ffi::c_void`
    --> /home/lxc/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libc-0.2.153/src/unix/mod.rs:1590:17
     |
1590 |         pub use ::ffi::c_void;
     |                 ^^^^^^^^^^^^^

Some errors have detailed explanations: E0432, E0463.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `libc` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
note: If the build failed due to a missing target, you can run this command:
note: 
note:     rustup target install aarch64-linux-android

**Comment by glime-ai** [comment]
Based on the error messages you've encountered, it seems like the root cause is that the Rust compiler cannot find the standard library for the `aarch64-linux-android` target. This typically happens when the target platform is not installed in your Rust toolchain. The error message even suggests a solution: to download the target using `rustup target add aarch64-linux-android`.

To resolve this issue, you should follow these steps:

1. Ensure that you have the Rust toolchain installed. If not, you can install it from [https://www.rust-lang.org/tools/install](https://www.rust-lang.org/tools/install).
2. Add the `aarch64-linux-android` target to your Rust toolchain by running the command suggested in the error message:
   ```
   rustup target add aarch64-linux-android
   ```
3. Once the target is added, try running your build command again:
   ```
   cargo ndk -t arm64-v8a -p 21 -- build --features ffi --release
   ```

This should resolve the errors related to not finding the `core` and `compiler_builtins` crates, as well as the unresolved import of `ffi::c_void`.

Additionally, ensure that you have the Android NDK installed and the `ANDROID_NDK_HOME` environment variable set correctly, as mentioned in the documentation. This is crucial for building projects targeting Android.

If you encounter further issues or the problem persists, consider checking the documentation for any additional prerequisites or steps that might have been missed, and ensure that all dependencies and tools are up to date.

**Reply by stopandjoin** [reply]
以上工具都安装过了，问题没有解决。

**Comment by stopandjoin** [comment]
求助一下 @iyangsj

**Reply by iyangsj** [reply, marked as answer]
注意使用rust官方方式来安装rust工具链，不要直接使用ubuntu的包管理工具来安装

```
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
rustup update
```

**Reply by iyangsj** [reply]
类似问题：https://github.com/rust-lang/rustup/issues/3426
